### PR TITLE
spatial/vptree: new package for vantage point tree NN search

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
This implements the optimised vp-tree algorithm from Yianilos, which says "An optimized tree results because function Select_vp endevours to pick better than random vantage points. Replacing it with a simple random selection generates a valid and effective tree; but experiments have shown that some effort towards making better choices, results in non-trivial search-time savings."

I don't see that.

```
Benchmark/NearestBruteN10:0-8         	     200	   6999116 ns/op
Benchmark/NearestBrute:0-8            	     300	   5947496 ns/op

Benchmark/Nearest:0-8                 	  300000	      5216 ns/op
Benchmark/Nearest:3-8                 	  300000	      5241 ns/op
Benchmark/Nearest:10-8                	  300000	      5173 ns/op
Benchmark/Nearest:30-8                	  300000	      5233 ns/op
Benchmark/Nearest:100-8               	  300000	      5258 ns/op
Benchmark/Nearest:300-8               	  300000	      5220 ns/op
Benchmark/NearestSetN10:0-8           	 1000000	      2207 ns/op
Benchmark/NearestSetN10:3-8           	 1000000	      2183 ns/op
Benchmark/NearestSetN10:10-8          	 1000000	      2151 ns/op
Benchmark/NearestSetN10:30-8          	 1000000	      2220 ns/op
Benchmark/NearestSetN10:100-8         	 1000000	      2176 ns/op
Benchmark/NearestSetN10:300-8         	 1000000	      2172 ns/op
```

With significant upfront costs for the additional work.
```
BenchmarkNew/New:0-8         	       5	 283258373 ns/op
BenchmarkNew/New:10-8        	       2	 875709497 ns/op
BenchmarkNew/New:100-8       	       1	7654383061 ns/op
```

Maybe the optimisation should be removed.

**Comment by kortschak** [comment]
The other question is whether there should be a user hook for injecting the random seed; should `New` also take a `src *rand.Source`?

File: spatial/vptree/vptree.go [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+	"sort"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/stat"
+)
+
+// Comparable is the element interface for values stored in a vp tree.
+type Comparable interface {
+	// Distance returns the Euclidean distance between the receiver and
```
Comment (vladimir-ch) [inline spatial/vptree/vptree.go]: If it's for general metric spaces, then the distance does not have to be Euclidean.

File: spatial/vptree/vptree.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+		effort = len(s)
+	}
+	var best Comparable
+	var bestVar float64
+	b.work = b.work[:effort*effort]
+	for i, p := range random(effort, s) {
+		for j, q := range random(effort, s) {
+			b.work[i*effort+j] = p.Distance(q)
+		}
+		variance := stat.Variance(b.work, nil)
```
Comment (vladimir-ch) [inline spatial/vptree/vptree.go]: I think this should use the i-th row slice of `work`.

File: spatial/vptree/vptree.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+	if n == nil {
+		return nil, inf
+	}
+
+	d := q.Distance(n.Point)
+	dist = math.Min(dist, d)
+
+	bn := n
+	if d < n.Radius {
+		if d-dist <= n.Radius {
```
Comment (vladimir-ch) [inline spatial/vptree/vptree.go]: This is equivalent to `if -dist <= n.Radius - d` but since `d < n.Radius` here, we are testing `if non-positive <= positive` which holds always. I find this suspicious.

File: spatial/vptree/vptree.go [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+		}
+		if d+dist >= n.Radius {
+			fn, fd := n.Further.search(q, dist)
+			if fd < dist {
+				dist = fd
+				bn = fn
+			}
+		}
+	} else {
+		if d+dist >= n.Radius {
```
Comment (vladimir-ch) [inline spatial/vptree/vptree.go]: Similarly here.

File: spatial/vptree/vptree.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+// Keep adds c to the heap if its distance is less than or equal to the max value of the heap.
+func (k *DistKeeper) Keep(c ComparableDist) {
+	if c.Dist <= k.Heap[0].Dist {
+		heap.Push(k, c)
+	}
+}
+
+// Keeper implements a conditional max heap sorted on the Dist field of the ComparableDist type.
+// kd search is guided by the distance stored in the max value of the heap.
```
Comment (vladimir-ch) [inline spatial/vptree/vptree.go]: What is `kd search` here? Should it be `vp search`?

File: spatial/vptree/vptree.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+	if n == nil {
+		return
+	}
+
+	k.Keep(ComparableDist{Comparable: n.Point, Dist: q.Distance(n.Point)})
+
+	d := q.Distance(n.Point)
+	dist := k.Max().Dist
+	if d < n.Radius {
+		if d-dist <= n.Radius {
```
Comment (vladimir-ch) [inline spatial/vptree/vptree.go]: Also here and below.

File: spatial/vptree/vptree_simple_example_test.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+		vptree.Point{7, 2},
+	}
+
+	// Print all points in the data set within 3 of (3, 5).
+	t := vptree.New(points, 0)
+	q := vptree.Point{3, 5}
+	t.Do(func(c vptree.Comparable, _ int) (done bool) {
+		// Compare each distance and output points
+		// with a Euclidean distance less than 3.
+		// Distance returns the square of the
```
Comment (vladimir-ch) [inline spatial/vptree/vptree_simple_example_test.go]: Copy/paste from `kdtree`? There is no square here.

File: spatial/vptree/vptree.go (hunk 1 above)
Comment (kortschak) [reply spatial/vptree/vptree.go]: Fixed

File: spatial/vptree/vptree.go (hunk 2 above)
Comment (kortschak) [reply spatial/vptree/vptree.go]: I'm an idiot.

File: spatial/vptree/vptree.go (hunk 5 above)
Comment (kortschak) [reply spatial/vptree/vptree.go]: Fixed

File: spatial/vptree/vptree_simple_example_test.go (hunk 7 above)
Comment (kortschak) [reply spatial/vptree/vptree_simple_example_test.go]: Fixed

File: spatial/vptree/vptree.go (hunk 3 above)
Comment (kortschak) [reply spatial/vptree/vptree.go]: I'll look into this further

**Comment by kortschak** [comment]
Now with the corrected selection calculation.
```
BenchmarkNew/New:0-8         	       5	 247633792 ns/op
BenchmarkNew/New:10-8        	       2	 770851644 ns/op
BenchmarkNew/New:100-8       	       1	3911120178 ns/op

Benchmark/NearestBruteN10:0-8         	     300	   5913136 ns/op
Benchmark/NearestBrute:0-8            	     300	   4935473 ns/op

Benchmark/Nearest:0-8                 	  300000	      4802 ns/op
Benchmark/Nearest:3-8                 	  300000	      4774 ns/op
Benchmark/Nearest:10-8                	  300000	      4590 ns/op
Benchmark/Nearest:30-8                	  300000	      4444 ns/op
Benchmark/Nearest:100-8               	  300000	      4366 ns/op
Benchmark/Nearest:300-8               	  300000	      4303 ns/op

Benchmark/NearestSetN10:0-8           	 1000000	      1971 ns/op
Benchmark/NearestSetN10:3-8           	 1000000	      1949 ns/op
Benchmark/NearestSetN10:10-8          	 1000000	      1971 ns/op
Benchmark/NearestSetN10:30-8          	 1000000	      1949 ns/op
Benchmark/NearestSetN10:100-8         	 1000000	      1988 ns/op
Benchmark/NearestSetN10:300-8         	 1000000	      1951 ns/op
```

File: spatial/vptree/vptree.go (hunk 3 above)
Comment (kortschak) [reply spatial/vptree/vptree.go]: Yes. The code came from algorithm 3 which includes the bounds which provide a better control of which branches to explore. The condition on the first option for each case can be omitted.

Comment (kortschak) [reply spatial/vptree/vptree.go]: Interestingly this condition is in all the reproductions of this algorithm.

1. [Yianilos](http://pnylab.com/papers/vptree/vptree.pdf)
2. [Fu _et al_](https://citeseerx.ist.psu.edu/viewdoc/download;jsessionid=477FE81DBEDC30B6A9240512C9CCF76D?doi=10.1.1.115.8253&rep=rep1&type=pdf).

and [this blog](http://stevehanov.ca/blog/index.php?id=130) and the remarkably similar [code here](https://github.com/lvdmaaten/bhtsne/blob/master/vptree.h#L229-L270).

File: spatial/vptree/vptree.go [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+// Comparable is the element interface for values stored in a vp-tree.
+type Comparable interface {
+	// Distance returns the distance between the receiver and the
+	// parameter. The returned distance must satisfy the properties
+	// of distances in a metric space.
+	//
+	// - a.Distance(a) == 0
+	// - a.Distance(b) >= 0
+	// - a.Distance(b) == b.Distance(a)
+	// - a.Distamce(b) <= a.Distamce(c)+c.Distance(b)
```
[the remark this answers is not in the document]
Comment (kortschak) [reply spatial/vptree/vptree.go]: Fixed

**Comment by kortschak** [comment]
Yeah. I guess it depends on how many times the tree will be used. I'm going to do some experiments to see when/if it pays off. Preliminarily, tree size makes makes a difference, but the benchmarks timeouted before I could see how much.

**Comment by kortschak** [comment]
It looks to me like it is extremely unlikely to pay off.

Here is a set of benchmarks with 1e4, 1e5 and 1e6 elements in the tree. For the 1e4 and 1e5 I have made unoptimised selection and perfectly optimised selection, for the 1e6 I have made unoptimised and optimisation with 100 samples because perfect optimisation is unfeasible at that scale.

We can see that for the 1e6 case, we would need to make ~5e6 queries to break even with the time investment, for 1e4 and 1e5 it is 15e4 to 14e6 and 6e5 to 5e8 queries respectively. In each case you're making more queries than there are elements in the tree.

There are some optimisations to make, but not really in the optimisation parts of the vp selection routines.

```
1e4 elements

BenchmarkNew/Newx10000:0-8                  	     100	  17065619 ns/op
BenchmarkNew/Newx10000:100-8                	      20	  60613299 ns/op
BenchmarkNew/Newx10000:10000-8              	       1	3153260301 ns/op
Benchmark/Nearestx10000:0-8                 	  500000	      2152 ns/op
Benchmark/Nearestx10000:100-8               	 1000000	      1870 ns/op
Benchmark/Nearestx10000:10000-8             	 1000000	      1926 ns/op
Benchmark/NearestSetN10x10000:0-8           	  100000	     15988 ns/op
Benchmark/NearestSetN10x10000:100-8         	  100000	     13808 ns/op
Benchmark/NearestSetN10x10000:10000-8       	  100000	     14567 ns/op

1e5 elements

BenchmarkNew/Newx100000:0-8                 	       5	 259904726 ns/op
BenchmarkNew/Newx100000:100-8               	       2	 682755984 ns/op
BenchmarkNew/Newx100000:100000-8                       1	595475559603 ns/op
Benchmark/Nearestx100000:0-8                	  200000	      5580 ns/op
Benchmark/Nearestx100000:100-8              	  300000	      4822 ns/op
Benchmark/Nearestx100000:100000-8           	  300000	      4489 ns/op
Benchmark/NearestSetN10x100000:0-8          	   50000	     36974 ns/op
Benchmark/NearestSetN10x100000:100-8        	   50000	     34467 ns/op
Benchmark/NearestSetN10x100000:100000-8     	   50000	     28500 ns/op

1e6 elements

BenchmarkNew/Newx1000000:0-8                       1	4480506641 ns/op
BenchmarkNew/Newx1000000:100-8                     1	8981545826 ns/op
Benchmark/Nearestx1000000:0-8                 200000	      9194 ns/op
Benchmark/Nearestx1000000:100-8               200000	      8378 ns/op
Benchmark/NearestSetN10x1000000:0-8            30000	     51336 ns/op
Benchmark/NearestSetN10x1000000:100-8          30000	     44197 ns/op
```

It's difficult to know whether this is expected, since Yianilos doesn't give any evidence in the paper for the performance benefits and with one exception (UnixJunkie/vp-tree) there is no implementation I can find that does the extra work (that is an OCaml implementation that also does the bounds in a way that is similar to vps). I have written to @UnixJunkie to see if he sees a worthwhile benefit.

I think that we should probably merge this as it is to keep the code in the history, but then back out the optimisation code, unless there is something I'm completely missing (also the rand source, since I think we can just always seed with 1 since there is no real benefit beyond repeatable execution to being able to specify the PRNG seed).

**Comment by UnixJunkie** [comment]
If you construct trees often, choosing vantage points randomly should be fine.
If you construct the tree once, but then query it many times, it might be worth choosing the VPs
in a smarter way.

Recently, I use bisector-trees instead of VP-trees in production, for static spatial indexing.
CF. https://github.com/UnixJunkie/bisec-tree
The "vantage points" selection heuristic is called two_bands in the code.
I think I saw it in a paper by Sergei Brin and called "greedy algorithm". It looks like a heuristic way
to find a double band of the point cloud (a diameter is always a double band, but the reverse is not true).
https://github.com/UnixJunkie/bisec-tree/blob/master/bisec_tree.ml
Choose one point randomly (call it A), then find the farthest from it (call it B), then the farthest from B (call it C).
Use B as your "vantage point" for the left sub-tree and C for the right sub-tree.

Note that benchmarks are just benchmarks, you would need point clouds from an actual problem
to be really sure (number of dimensions of the points, number of points, number of queries per constructed tree, actual distance function being used, etc).

**Comment by kortschak** [comment]
Thanks for the input. The trouble that I'm having is seeing room for an improvement (heeding caveat on benchmarks) that would pay for the cost of calculation of optimalish vantage points. The pay off (from random data) looks like it's query count of about 10x the number of points stored in the tree. This is an unlikely use case.

**Comment by UnixJunkie** [comment]
Some machine learning methods use nearest-neighbor searches a lot.
Indexing the training set well is really worthy in some situations.

**Comment by kortschak** [comment]
Thanks. What has your experience been with the bisector trees cf vp trees?

**Comment by UnixJunkie** [comment]
I don't have benchmarks, but:
i) my BST implementation is bucketized; so it is more advanced than my VPT implementation.
Bucket size higher than one allows the tree to use less memory, and buckets could
be fetched from disk on demand, for very large datasets.
Note that having buckets of size one is equivalent to a non bucketized BST.
ii) I did not understand in Yianilos' paper how to do that with VPTs (despites he explains and draws
an example).
iii) BSTs have a theoretical advantage in high dimensions: they split the space more equally at each step. Sorry, I don't remember where I read about this.

**Comment by vladimir-ch** [comment]
Any conclusion from this, @kortschak ?

**Comment by kortschak** [comment]
The conclusion in am email was that it can make a difference in some machine learning contexts and that we should leave it up to the user. On the basis of that, this is ready.
